src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -25,121 +25,113 @@ private sealed class ChunkedEncodingReadStream : HttpContentReadStream
+                {
+                    // Cancellation requested.
+                    return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
+                }
 
-                // Read the start of the chunk line.
-                _connection._allowedReadLineBytes = MaxChunkBytesAllowed;
-                ArraySegment<byte> line = await _connection.ReadNextLineAsync().ConfigureAwait(false);
+                if (_connection == null || destination.Length == 0)
+                {
+                    // Response body fully consumed or the caller didn't ask for any data.
+                    return new ValueTask<int>(0);
+                }
 
-                // Parse the hex value.
-                if (!Utf8Parser.TryParse(line.AsReadOnlySpan(), out ulong chunkSize, out int bytesConsumed, 'X'))
+                // Try to consume from data we already have in the buffer.
+                int bytesRead = ReadChunksFromConnectionBuffer(destination.Span);
+                if (bytesRead > 0)
                 {
-                    throw new IOException(SR.net_http_invalid_response);
+                    return new ValueTask<int>(bytesRead);
                 }
-                else if (bytesConsumed != line.Count)
+
+                // Nothing available to consume.  Fall back to I/O.
+                return ReadAsyncCore(destination, cancellationToken);
+            }
+
+            private async ValueTask<int> ReadAsyncCore(Memory<byte> destination, CancellationToken cancellationToken)
+            {
+                // Should only be called if ReadChunksFromConnectionBuffer returned 0.
+
+                Debug.Assert(_connection != null);
+                Debug.Assert(destination.Length > 0);
+
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
+                try
                 {
-                    // There was data after the chunk size, presumably a "chunk extension".
-                    // Allow tabs and spaces and then stop validating once we get to an extension.
-                    int offset = line.Offset + bytesConsumed, end = line.Count - bytesConsumed + line.Offset;
-                    for (int i = offset; i < end; i++)
+                    while (true)
                     {
-                        char c = (char)line.Array[i];
-                        if (c == ';')
+                        if (_state == ParsingState.Ending)
                         {
-                            break;
+                            await ConsumeTrailingHeaders().ConfigureAwait(false);
+                            Finish();
+                            return 0;
                         }
-                        else if (c != ' ' && c != '\t') // not called out in the RFC, but WinHTTP allows it
+
+                        // We're only here if we need more data to make forward progress.
+                        await _connection.FillAsync();
```

Comment:
If I understand this properly, we are always doing buffered reads now, right?

It still seems valuable to me to do an unbuffered read in at least some situations.

Reply:
I'll add that.  To be clear, you mean adding a check above this like the following, right?
```C#
if (_state == ParsingState.ExpectChunkData &&
    (ulong)destination.Length > _chunkBytesRemaining)
{
    int bytesRead = await _connection.ReadAsync(destination.Slice(0, (int)_chunkBytesRemaining));
    if (bytesRead == 0) throw new IOException(SR.net_http_invalid_response);
    return bytesRead;
}
```

Reply:
Yes, though I think it's destinationLength <= _chunkBytesRemaining

Reply:
And you may need to change state if you read to the end of the chunk

Reply:
nevermind, my comment about the compare is wrong...